src/instrumentsscene/view/instrumentlistmodel.cpp

```diff
@@ -272,373 +256,233 @@ void InstrumentListModel::sortInstruments(QVariantList& instruments) const
+        setCurrentGenre(genres[index]->id);
     }
-
-    m_selectedGroupId = groupId;
-
-    emit dataChanged();
-    emit selectedGroupChanged(groupId);
 }
 
-void InstrumentListModel::selectInstrument(const QString& instrumentName, const QString& traitName)
+void InstrumentListModel::setCurrentGroupIndex(int index)
 {
-    const InstrumentTemplate* suitedTemplate = nullptr;
-
-    for (const InstrumentTemplate* templ : m_instrumentsMeta.instrumentTemplates) {
-        if (templ->trackName == instrumentName && templ->trait.name == traitName) {
-            suitedTemplate = templ;
-            break;
-        }
-    }
-
-    if (!suitedTemplate) {
-        LOGE() << QString("Instrument %1 with trait %2 does not exist")
-            .arg(instrumentName)
-            .arg(traitName);
-        return;
-    }
+    InstrumentGroupList groups = availableGroups();
 
-    SelectedInstrumentInfo info;
-    info.isExistingPart = false;
-    info.isSoloist = false;
-    info.id = suitedTemplate->id;
-    info.name = formatInstrumentTitle(suitedTemplate->trackName, suitedTemplate->trait);
-    info.familyCode = suitedTemplate->familyId();
-    info.config = *suitedTemplate;
-
-    if (!m_canSelectMultipleInstruments) {
-        m_selectedInstruments.clear();
-        m_selectedInstruments << info;
-    } else if (!m_scoreOrders[m_selectedScoreOrderIndex].info.isValid()) {
-        m_selectedInstruments << info;
-    } else {
-        m_selectedInstruments.insert(instrumentInsertIndex(info), info);
+    if (index >= 0 && index < groups.size()) {
+        setCurrentGroup(groups[index]->id);
     }
-
-    emit selectedInstrumentsChanged();
 }
 
-void InstrumentListModel::unselectInstrument(int instrumentIndex)
+void InstrumentListModel::selectInstrument(int instrumentIndex)
 {
     if (!isInstrumentIndexValid(instrumentIndex)) {
         return;
     }
 
-    m_selectedInstruments.removeAt(instrumentIndex);
-    emit selectedInstrumentsChanged();
-}
+    QModelIndex modelIndex = index(instrumentIndex);
+    m_selection->select(modelIndex);
 
-void InstrumentListModel::swapSelectedInstruments(int firstInstrumentIndex, int secondInstrumentIndex)
-{
-    m_selectedInstruments.swapItemsAt(firstInstrumentIndex, secondInstrumentIndex);
-    emit selectedInstrumentsChanged();
-    checkScoreOrderMatching(true);
-}
+    if (isSearching()) {
+        const CombinedInstrument& instrument = m_instruments[instrumentIndex];
 
-void InstrumentListModel::setSearchText(const QString& text)
-{
-    if (m_searchText == text) {
-        return;
+        if (m_selection->selection().size() == 1 && !instrument.templates.isEmpty()) {
+            doSetCurrentGroup(instrument.templates.first()->groupId);
+        } else {
+            doSetCurrentGroup(NONE_GROUP_ID);
+        }
     }
 
-    m_searchText = text;
-    emit dataChanged();
-
-    updateGenreStateBySearch();
+    emit dataChanged(index(0), index(rowCount() - 1), { RoleIsSelected });
```

Comment:
Can you notify about the selected state changed only for the current instrument(modelIndex)?